sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala

```diff
@@ -0,0 +1,151 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{Alias, Ascending, Attribute, AttributeReference, Cast, SortOrder}
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan, Project, Sort}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreePattern
+import org.apache.spark.sql.types.{AtomicType, DataType, StringType}
+import org.apache.spark.unsafe.types.UTF8String
+
+
+class ResolveTranspose(sparkSession: SparkSession) extends Rule[LogicalPlan] {
+
+  private def leastCommonType(dataTypes: Seq[DataType]): DataType = {
+    if (dataTypes.isEmpty) {
+      StringType
+    } else {
+      dataTypes.reduce { (dt1, dt2) =>
+        TypeCoercion.findTightestCommonType(dt1, dt2).getOrElse {
+          throw new IllegalArgumentException(s"No common type found for $dt1 and $dt2")
+        }
+      }
+    }
+  }
+
+  private def transposeMatrix(
+      fullCollectedRows: Array[InternalRow],
+      nonIndexColumnNames: Seq[String],
+      nonIndexColumnDataTypes: Seq[DataType]): Array[Array[Any]] = {
+    // Construct the original matrix
+    val originalMatrixNumCols = fullCollectedRows.head.numFields - 1
+    val originalMatrixNumRows = fullCollectedRows.length
+
+    val originalMatrix = Array.ofDim[Any](originalMatrixNumRows, originalMatrixNumCols)
+    for (i <- 0 until originalMatrixNumRows) {
+      val row = fullCollectedRows(i)
+      for (j <- 0 until originalMatrixNumCols) {
+        originalMatrix(i)(j) = row.get(j + 1, nonIndexColumnDataTypes(j))
+      }
+    }
+
+    // Transpose the original matrix
+    val transposedMatrix = Array.ofDim[Any](originalMatrixNumCols, originalMatrixNumRows)
+    for (i <- 0 until originalMatrixNumRows) {
+      for (j <- 0 until originalMatrixNumCols) {
+        transposedMatrix(j)(i) = originalMatrix(i)(j)
+      }
+    }
+
+    // Insert nonIndexColumnNames as first "column"
+    val finalMatrix = Array.ofDim[Any](originalMatrixNumCols, originalMatrixNumRows + 1)
+    for (i <- 0 until originalMatrixNumCols) {
+      finalMatrix(i)(0) = UTF8String.fromString(nonIndexColumnNames(i))
+    }
+    for (i <- 0 until originalMatrixNumCols) {
+      for (j <- 1 until originalMatrixNumRows + 1) {
+        finalMatrix(i)(j) = transposedMatrix(i)(j - 1)
+      }
+    }
+
+    finalMatrix
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(TreePattern.UNRESOLVED_TRANSPOSE)) {
+    case t @ UnresolvedTranspose(indexColumn, child, _, _) if !t.resolved =>
```

Comment:
```suggestion
    case t @ UnresolvedTranspose(indexColumn, child, _, _) if indexColumn.resolved && child.resolved =>
```

Reply:
Adjusted.

May I ask why not `!t.resolve`?